File: src/test/java/io/pivotal/kotlindemo/java/clients/ClientRepositoryTest.java
package io.pivotal.kotlindemo.java.clients;

import org.junit.Before;
import org.junit.Test;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

public abstract class ClientRepositoryTest {
    public abstract ClientRepository getClientRepository();

    private ClientRepository clientRepository;

    @Before
    public void setup() {
        clientRepository = getClientRepository();
    }

    @Test
    public void create_returnsANewClientWithAUniqueId() {
        CreateClientRequest createClientRequest1 = new CreateClientRequest(
                "Alex",
                "Basson",
                "abasson@pivotal.io"
        );

        Client client1 = clientRepository.create(createClientRequest1);

        assertThat(client1.getId()).isNotNull();
        assertThat(client1).isEqualToIgnoringGivenFields(
                new Client(
                        null,
                        createClientRequest1.getFirstName(),
                        createClientRequest1.getLastName(),
                        createClientRequest1.getEmail()
                ),
                "id"
        );

        CreateClientRequest createClientRequest2 = new CreateClientRequest(
                "David",
                "Edwards",
                "dedwards@pivotal.io"
        );

        Client client2 = clientRepository.create(createClientRequest2);

        assertThat(client2.getId()).isNotNull();
        assertThat(client2).isEqualToIgnoringGivenFields(
                new Client(
                        null,
                        createClientRequest2.getFirstName(),
                        createClientRequest2.getLastName(),
                        createClientRequest2.getEmail()
                ),
                "id"
        );

        assertThat(client1.getId()).isNotEqualTo(client2.getId());
    }

    @Test
    public void findAll_returnsAllClients() {
        CreateClientRequest createClientRequest1 = new CreateClientRequest(
                "Alex",
                "Basson",
                "abasson@pivotal.io"
        );

        Client client1 = clientRepository.create(createClientRequest1);

        CreateClientRequest createClientRequest2 = new CreateClientRequest(
                "David",
                "Edwards",
                "dedwards@pivotal.io"
        );

        Client client2 = clientRepository.create(createClientRequest2);

        assertThat(clientRepository.findAll()).containsExactlyInAnyOrder(client1, client2);
    }

    @Test
    public void findById_whenAClientWithTheGivenIdExists_returnsTheClientWithTheGivenId() {
        CreateClientRequest createClientRequest1 = new CreateClientRequest(
                "Alex",
                "Basson",
                "abasson@pivotal.io"
        );

        Client client1 = clientRepository.create(createClientRequest1);

        CreateClientRequest createClientRequest2 = new CreateClientRequest(
                "David",
                "Edwards",
                "dedwards@pivotal.io"
        );

        Client client2 = clientRepository.create(createClientRequest2);

        assertThat(clientRepository.findById(client1.getId())).isEqualTo(client1);
        assertThat(clientRepository.findById(client2.getId())).isEqualTo(client2);
    }

    @Test(expected = NoClientFoundException.class)
    public void findById_whenNoClientWithTheGivenIdExists_throwsANoClientFoundException() {
        Long nonExistantClientId = 12345L;
        clientRepository.findById(nonExistantClientId);
    }

    @Test
    public void findByLastName_returnsAPossiblyEmptyListOfClientsWithTheGivenLastName() {
        CreateClientRequest createClientRequest1 = new CreateClientRequest(
                "Alex",
                "Basson",
                "abasson@pivotal.io"
        );

        Client alexBasson = clientRepository.create(createClientRequest1);

        CreateClientRequest createClientRequest2 = new CreateClientRequest(
                "David",
                "Edwards",
                "dedwards@pivotal.io"
        );

        Client davidEdwards = clientRepository.create(createClientRequest2);

        CreateClientRequest createClientRequest3 = new CreateClientRequest(
                "Phoebe",
                "Basson",
                "abasson@pivotal.io"
        );

        Client phoebeBasson = clientRepository.create(createClientRequest3);

        assertThat(clientRepository.findByLastName("Basson")).containsExactlyInAnyOrder(
                alexBasson,
                phoebeBasson
        );
        assertThat(clientRepository.findByLastName("Edwards")).containsExactlyInAnyOrder(
                davidEdwards
        );
        assertThat(clientRepository.findByLastName("NonExistant")).isEmpty();
    }

    @Test
    public void findByEmail_whenAClientWithTheGivenEmailExists_returnsAnOptionalClientWithTheGivenEmail() {
        CreateClientRequest createClientRequest1 = new CreateClientRequest(
                "Alex",
                "Basson",
                "abasson@pivotal.io"
        );

        Client alexBasson = clientRepository.create(createClientRequest1);

        CreateClientRequest createClientRequest2 = new CreateClientRequest(
                "David",
                "Edwards",
                "dedwards@pivotal.io"
        );

        Client davidEdwards = clientRepository.create(createClientRequest2);

        assertThat(clientRepository.findByEmail(alexBasson.getEmail())).isEqualTo(Optional.of(alexBasson));
        assertThat(clientRepository.findByEmail(davidEdwards.getEmail())).isEqualTo(Optional.of(davidEdwards));
    }

    @Test
    public void findByEmail_whenNoClientWithTheGivenEmailExists_returnsAnEmptyOptional() {
        assertThat(clientRepository.findByEmail("no.such.client@email.com").isPresent()).isFalse();
    }

}

File: src/test/kotlin/io/pivotal/kotlindemo/kotlin/clients/ClientRepositoryTest.kt
package io.pivotal.kotlindemo.kotlin.clients

import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test

abstract class ClientRepositoryTest {
    abstract fun getClientRepository(): ClientRepository

    private lateinit var clientRepository: ClientRepository

    @Before
    fun setup() {
        clientRepository = getClientRepository()
    }

    @Test
    fun `create() returns a new client with a unique id`() {
        val createClientRequest1 = buildCreateClientRequest()

        val client1 = clientRepository.create(createClientRequest1)

        val createClientRequest2 = buildCreateClientRequest()

        assertThat(client1).isEqualToIgnoringGivenFields(
            buildClient(
                firstName = createClientRequest1.firstName,
                lastName = createClientRequest1.lastName,
                email = createClientRequest1.email
            ),
            "id"
        )

        val client2 = clientRepository.create(createClientRequest2)

        assertThat(client2).isEqualToIgnoringGivenFields(
            buildClient(
                firstName = createClientRequest2.firstName,
                lastName = createClientRequest2.lastName,
                email = createClientRequest2.email
            ),
            "id"
        )

        assertThat(client1.id).isNotEqualTo(client2.id)
    }

    @Test
    fun `findAll() returns all clients`() {
        val client1 = clientRepository.create(buildCreateClientRequest())
        val client2 = clientRepository.create(buildCreateClientRequest())

        assertThat(clientRepository.findAll()).containsExactlyInAnyOrder(client1, client2)
    }

    @Test
    fun `findById(), when a client with the given id exists, returns the client with the given id`() {
        val client1 = clientRepository.create(buildCreateClientRequest())
        val client2 = clientRepository.create(buildCreateClientRequest())

        assertThat(clientRepository.findById(client1.id)).isEqualTo(client1)
        assertThat(clientRepository.findById(client2.id)).isEqualTo(client2)
    }

    @Test(expected = NoClientFoundException::class)
    fun `findById(), when no client with the given id exists, throws a NoClientFoundException`() {
        val nonExistantClientId = 12345L
        clientRepository.findById(nonExistantClientId)
    }

    @Test
    fun `findByLastName() returns a possibly empty list of clients with the given last name`() {
        val alexBasson = clientRepository.create(buildCreateClientRequest(lastName = "Basson"))
        val phoebeBasson = clientRepository.create(buildCreateClientRequest(lastName = "Basson"))
        val davidEdwards = clientRepository.create(buildCreateClientRequest(lastName = "Edwards"))

        assertThat(clientRepository.findByLastName("Basson")).containsExactlyInAnyOrder(
            alexBasson,
            phoebeBasson
        )
        assertThat(clientRepository.findByLastName("Edwards")).containsExactlyInAnyOrder(
            davidEdwards
        )
        assertThat(clientRepository.findByLastName("NonExistant")).isEmpty()
    }

    @Test
    fun `findByEmail(), when a client with the given email exists, returns the client with given email`() {
        val alexBasson = clientRepository.create(buildCreateClientRequest(email = "abasson@pivotal.io"))
        val davidEdwards = clientRepository.create(buildCreateClientRequest(email = "dedwards@pivotal.io"))

        assertThat(clientRepository.findByEmail(alexBasson.email)).isEqualTo(alexBasson)
        assertThat(clientRepository.findByEmail(davidEdwards.email)).isEqualTo(davidEdwards)
    }

    @Test
    fun `findByEmail(), when no client with the given email exists, returns null`() {
        assertThat(clientRepository.findByEmail("no.such.client@email.com")).isNull()
    }

}
